### ISSUE TITLE
Crashes on multiple site (related to Service Workers)

### ISSUE THREAD

**Issue by jhhuisman** [qutebrowser/qutebrowser#5279]
**Version info**:
qutebrowser v1.10.1
Git commit: df45e7401 (2020-02-22 10:27:28 +0100)
Backend: QtWebEngine (Chromium 77.0.3865.129)
Qt: 5.14.1

CPython: 3.6.9
PyQt: 5.14.1

sip: 5.1.1
colorama: 0.4.3
pypeg2: 2.15
jinja2: 2.11.1
pygments: 2.5.2
yaml: 5.3
cssutils: 1.0.2 $Id$
attr: 19.3.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.14.0
PyQt5.QtWebKitWidgets: no
pdf.js: no
sqlite: 3.30.1
QtNetwork SSL: OpenSSL 1.1.1d  10 Sep 2019

Style: QFusionStyle
Platform: Linux-4.15.0-88-generic-x86_64-with-Ubuntu-18.04-bionic, 64bit
Linux distribution: Ubuntu 18.04.4 LTS (ubuntu)

**Does the bug happen if you start with `--temp-basedir`?**:
No, it does not crash with --temp-basedir.

What does this mean?
Something in my config?
What info do you need?

**Description**
I love Qutebrowser and I want to replace my old browser.
However it crashes on some sites: Reddit, Teamleader and Toggl app and probably others.
Do you have any suggestions/fixes? Thanks so much.

**How to reproduce**
This page crashes the browser for example:
https://www.reddit.com/r/qutebrowser/comments/9wvgv2/qutebrowser_crashes/

This one also:
https://toggl.com/app/timer

This is the error:
==== Fault log ====
Fatal Python error: Segmentation fault

Thread 0x00007ff071358740 (most recent call first):
  File "/home/jacob/qutebrowser/qutebrowser/app.py", line 134 in qt_mainloop
  File "/home/jacob/qutebrowser/qutebrowser/app.py", line 124 in run
  File "/home/jacob/qutebrowser/qutebrowser/qutebrowser.py", line 201 in main
  File "/home/jacob/qutebrowser/qutebrowser/__main__.py", line 29 in <module>
  File "/usr/lib/python3.6/runpy.py", line 85 in _run_code
  File "/usr/lib/python3.6/runpy.py", line 193 in _run_module_as_main

**Comment by jhhuisman** [comment]
I seem to have solved this issue by following tips from other issues.
Deleted ~/.local/share/qutebrowser/webengine/Service\ Worker

https://github.com/qutebrowser/qutebrowser/issues/4851
https://github.com/qutebrowser/qutebrowser/issues/4853

**Comment by The-Compiler** [comment]
qutebrowser v1.10.1 should do that automatically on Qt updates - do you have a `.bak` folder there still? I've never been able to reproduce this (and neither has Qt where I reported this upstream), so I'd really need some data from there to be able to say more.

**Comment by jhhuisman** [comment]
This morning it crashed again. So it might not have helped.
The Service Worker dir is back also.
I zipped it and also the -bak folder.
How can I send those to you?

**Comment by The-Compiler** [comment]
Meh... I'll probably get back to you next week to test Qt 5.14.2 and Qt 5.15 Beta 2, in the hope that something improved there.

You can reach me via mail@qutebrowser.org - I'm not sure if the Service Workers dir only will be enough, but I understand if you don't want to send your entire browsing data and stuff.

If you find the time, maybe it'd be possible for you to reproduce this in a new session with `--basedir testdir` without providing any personal data? Then you could send that entire directory to me, which would hopefully allow me (and Qt developers) to reproduce this properly. You'd probably need to visit some affected websites and maybe quit/restart a couple of times, in the hope that it comes back.

Either way, thanks for the report!

**Comment by jhhuisman** [comment]
It doesn't seem to crash anymore. The websites I listed before work fine now.
I guess it had to do with that I installed Qutebrowser twice.
https://github.com/qutebrowser/qutebrowser/issues/5272

Let me know if you still need anything.

**Comment by schyzophrene-asynchrone** [comment]
As far as I’m concerned unity.com is a website that always causes qutebrowser to crash. I’ve just tested it with `--basedir testdir` (so it shouldn’t be affected by my personal config) and it still crashes.
[testdir.tar.gz](https://github.com/qutebrowser/qutebrowser/files/4387767/testdir.tar.gz)
Let me know if you need more information, as I don’t really know what to send to you.

**Comment by The-Compiler** [comment]
@schyzophrene-asynchrone No issues here. Can you show your version information (`:version`)? Also, what OS/distribution are you on?

**Comment by schyzophrene-asynchrone** [comment]
`qutebrowser v1.10.1
Git commit: 
Backend: QtWebEngine (Chromium 77.0.3865.129)
Qt: 5.14.1

CPython: 3.8.2
PyQt: 5.14.1

sip: 5.1.1
colorama: 0.4.3
pypeg2: 2.15
jinja2: 2.11.1
pygments: 2.6.1
yaml: 5.3
cssutils: no
attr: 19.3.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.14.0
PyQt5.QtWebKitWidgets: yes
pdf.js: no
sqlite: 3.31.1
QtNetwork SSL: OpenSSL 1.1.1e  17 Mar 2020

Style: QFusionStyle
Platform: Linux-5.5.13-arch1-1-x86_64-with-glibc2.2.5, 64bit
Linux distribution: Arch Linux (arch)
Frozen: False
Imported from /usr/lib/python3.8/site-packages/qutebrowser
Using Python from /usr/bin/python3
Qt library executable path: /usr/lib/qt/libexec, data path: /usr/share/qt

Paths:
cache: /home/hoel/testdir/cache
config: /home/hoel/testdir/config
data: /home/hoel/testdir/data
runtime: /home/hoel/testdir/runtime
system data: /usr/share/qutebrowser

Autoconfig loaded: yes
Config.py: no config.py was loaded
Uptime: 0:00:09`
As you can see, I’m running Arch Linux, freshly updated.

**Comment by The-Compiler** [comment]
@schyzophrene-asynchrone Odd. Can you reproduce the same issue in KDE Falkon (package `falkon`) as well? I'll get back to you once Qt 5.14.2 is released (had a bit of a delay) and my debug packages are updated.

**Comment by schyzophrene-asynchrone** [comment]
Yes, exactly the same : I go to unity.com, wait a bit (till the tab title changes actually) and I get a coredump.

**Comment by schyzophrene-asynchrone** [comment]
On a side note (don’t know wether that might be important or not), I’m running sway under wayland (with xwayland enabled). That happened to be a relevant information while debugging some apps using SDL.

**Comment by The-Compiler** [comment]
@schyzophrene-asynchrone Okay, at this point I'm pretty sure this issue is unrelated to what's described here. Could you open a new issue with the information above please, to make it easier to me to keep track of everything? Also, can you please try to [get a stacktrace](https://github.com/qutebrowser/qutebrowser/blob/master/doc/stacktrace.asciidoc#getting-the-stack-trace) as well? Ignore the part about installing debug packages, they are currently outdated because the hackerspace hosting the build server is moving.

**Comment by bitraid** [comment]
This might be a wayland issue (too). @schyzophrene-asynchrone are you using wayland? If so, can you try with XWayland?

**Comment by bitraid** [comment]
@schyzophrene-asynchrone  Does it crash if you start qutebrowser with `QT_QPA_PLATFORM= qutebrowser` ?

**Comment by schyzophrene-asynchrone** [comment]
No it doesn’t. I created a new issue with all related informations (#5313).

**Comment by The-Compiler** [comment]
Apparently Qt hasn't been able to reproduce it with Qt 5.14.1: https://bugreports.qt.io/browse/QTBUG-82105

So we're back to square one with this... Anyone has tested whether things improved with 5.14.2 yet?

**Comment by The-Compiler** [comment]
Just got confirmation that this happens with Qt 5.14.2 as well, when visiting Outlook Web.

**Comment by Aledosim** [comment]
Hello there.
As I mentioned in #5423 it happened again. After moved "webengine/Service Worker" directory in qutebrowser's data dir away it worked

**Comment by The-Compiler** [comment]
Anyone knows whether `--qt-flag disable-shared-workers` makes any difference?

**Comment by sentriz** [comment]
@The-Compiler not for me (is it just `$ qutebrowser --qt-flag  disabled-shared-workers`?) `old.reddit.com` is crashing

**Comment by The-Compiler** [comment]
@sentriz Yep - I don't expect it to make a difference to be honest, but it was worth a try. Please make sure you don't get a crash with `--temp-basedir` though, as there are some other things going on as well (notably a bug in mesa causing crashes).

**Comment by sentriz** [comment]
yeah same thing with `$ qutebrowser --temp-basedir --qt-flag disabled-shared-workers` maybe I can try a specific qt-webengine version

**Comment by The-Compiler** [comment]
@sentriz You're not seeing this issue then. If you're on Archlinux, you might need to downgrade mesa - see #5462.

**Comment by The-Compiler** [comment]
I've just seen this happen with Qt 5.15 when running simplebrowser - however, I'm not sure if that's the earlier issue where the crashes happen on Qt upgrades only (which was the case here), or if that's related to the issue which seems to keep coming back. At least I now have some data I might be able to share with Qt after taking a closer look at what's saved in there (note to self: `cp -R ~/.local/share/QtExamples/simplebrowser/QtWebEngine/Default ~/proj/qutebrowser/issue-5279-simplebrowser-webengine`)

**Comment by The-Compiler** [comment]
I *think* I finally found a way to consistently reproduce this (though I don't have a QtWebEngine build with debugging symbols ready at the moment), by opening the Epic Games store:

- Open https://www.epicgames.com/store/en-US/product/old-world/home with a fresh basedir (works fine)
- Quit
- Open it again
- Crash in `Chrome_IOThread`

**Comment by toofar** [comment]
I have a debug build for 5.14 if that helps (and the reproducer you described works there):
<details>
  <summary>backtrace</summary>

```
mojo::Message::SerializeAssociatedEndpointHandles ()
    at ../../../../../qtwebengine/src/3rdparty/chromium/mojo/public/cpp/bindings/lib/message.cc:464
464	        group_controller->AssociateInterface(std::move(handle));
(gdb) bt
#0  mojo::Message::SerializeAssociatedEndpointHandles ()
    at ../../../../../qtwebengine/src/3rdparty/chromium/mojo/public/cpp/bindings/lib/message.cc:464
#1  0x00007fffebf62d77 in mojo::InterfaceEndpointClient::SendMessage
    ()
    at ../../../../../qtwebengine/src/3rdparty/chromium/mojo/public/cpp/bindings/lib/interface_endpoint_client.cc:264
#2  0x00007fffea580e27 in blink::mojom::ServiceWorkerContainerProxy::SetController ()
    at ./gen/third_party/blink/public/mojom/service_worker/service_worker_container.mojom.cc:1792
#3  0x00007fffeef4c28d in content::ServiceWorkerProviderHost::SendSetControllerServiceWorker ()
    at ../../../../../qtwebengine/src/3rdparty/chromium/content/browser/service_worker/service_worker_provider_host.cc:799
#4  0x00007fffef2b2ab6 in content::ServiceWorkerControlleeRequestHandler::ContinueWithActivatedVersion ()
    at ../../../../../qtwebengine/src/3rdparty/chromium/content/browser/service_worker/service_worker_controllee_request_handler.cc:370
#5  0x00007fffef2b3758 in content::ServiceWorkerControlleeRequestHandler::ContinueWithRegistration ()
    at ../../../../../qtwebengine/src/3rdparty/chromium/content/browser/service_worker/service_worker_controllee_request_handler.cc:322
#6  0x00007fffef2b193b in base::internal::FunctorTraits<void (content::ServiceWorkerControlleeRequestHandler::*)(blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>), void>::Invoke<void (content::ServiceWorkerControlleeRequestHandler::*)(blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>), base::WeakPtr<content::ServiceWorkerControlleeRequestHandler>, blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration> > ()
    at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:499
#7  base::internal::InvokeHelper<true, void>::MakeItSo<void (content::ServiceWorkerControlleeRequestHandler::*)(blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>), base::WeakPtr<content::ServiceWorkerControlleeRequestHandler>, blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration> > ()
    at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:619
#8  base::internal::Invoker<base::internal::BindState<void (content::ServiceWorkerControlleeRequestHandler::*)(blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>), base::WeakPtr<content::ServiceWorkerControlleeRequestHandler> >, void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>::RunImpl<void (content::ServiceWorkerControlleeRequestHandler::*)(blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>), std::tuple<base::WeakPtr<content::ServiceWorkerControlleeRequestHandler> >, 0ul>(void (content::ServiceWorkerControlleeRequestHandler::*&&)(blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>), std::tuple<base::WeakPtr<content::ServiceWorkerControlleeRequestHandler> >&&, std::integer_sequence<unsigned long, 0ul>, blink::ServiceWorkerStatusCode&&, scoped_refptr<content::ServiceWorkerRegistration>&&) ()
    at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:673
#9  base::internal::Invoker<base::internal::BindState<void (content::ServiceWorkerControlleeRequestHandler::*)(blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>), base::WeakPtr<content::ServiceWorkerControlleeRequestHandler> >, void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>::RunOnce(base::internal::BindStateBase*, blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>&&) () at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:641
#10 0x00007fffeef5fc21 in base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>::Run(blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>) && () at ../../../../../qtwebengine/src/3rdparty/chromium/base/callback.h:98
#11 CompleteFindNow () at ../../../../../qtwebengine/src/3rdparty/chromium/content/browser/service_worker/service_worker_storage.cc:53
#12 0x00007fffeef6c8ff in content::ServiceWorkerStorage::ReturnFoundRegistration(base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&) () at ../../../../../qtwebengine/src/3rdparty/chromium/content/browser/service_worker/service_worker_storage.cc:1335
#13 0x00007fffeef6ca48 in content::ServiceWorkerStorage::DidFindRegistrationForDocument(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status) () at ../../../../../qtwebengine/src/3rdparty/chromium/content/browser/service_worker/service_worker_storage.cc:1239
#14 0x00007fffeef5c793 in base::internal::FunctorTraits<void (content::ServiceWorkerStorage::*)(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status), void>::Invoke<void (content::ServiceWorkerStorage::*)(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status), base::WeakPtr<content::ServiceWorkerStorage>, GURL, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status>(void (content::ServiceWorkerStorage::*)(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status), base::WeakPtr<content::ServiceWorkerStorage>&&, GURL&&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>&&, long&&, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status&&) () at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:499
#15 base::internal::InvokeHelper<true, void>::MakeItSo<void (content::ServiceWorkerStorage::*)(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status), base::WeakPtr<content::ServiceWorkerStorage>, GURL, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status>(void (content::ServiceWorkerStorage::*&&)(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status), base::WeakPtr<content::ServiceWorkerStorage>&&, GURL&&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>&&, long&&, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status&&) () at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:619
#16 base::internal::Invoker<base::internal::BindState<void (content::ServiceWorkerStorage::*)(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status), base::WeakPtr<content::ServiceWorkerStorage>, GURL, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long>, void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>::RunImpl<void (content::ServiceWorkerStorage::*)(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status), std::tuple<base::WeakPtr<content::ServiceWorkerStorage>, GURL, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long>, 0ul, 1ul, 2ul, 3ul>(void (content::ServiceWorkerStorage::*&&)(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status), std::tuple<base::WeakPtr<content::ServiceWorkerStorage>, GURL, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long>&&, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul, 3ul>, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status&&) () at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:673
#17 base::internal::Invoker<base::internal::BindState<void (content::ServiceWorkerStorage::*)(GURL const&, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status), base::WeakPtr<content::ServiceWorkerStorage>, GURL, base::OnceCallback<void (blink::ServiceWorkerStatusCode, scoped_refptr<content::ServiceWorkerRegistration>)>, long>, void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>::RunOnce(base::internal::BindStateBase*, content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status) () at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:641
#18 0x00007fffeef5d9a9 in base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>::Run(content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status) && () at ../../../../../qtwebengine/src/3rdparty/chromium/base/callback.h:98
#19 base::internal::FunctorTraits<base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>, void>::Invoke<base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>, content::ServiceWorkerDatabase::RegistrationData, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> >, content::ServiceWorkerDatabase::Status>(base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>&&, content::ServiceWorkerDatabase::RegistrationData&&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> >&&, content::ServiceWorkerDatabase::Status&&) () at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:561
#20 base::internal::InvokeHelper<false, void>::MakeItSo<base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>, content::ServiceWorkerDatabase::RegistrationData, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> >, content::ServiceWorkerDatabase::Status>(base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>&&, content::ServiceWorkerDatabase::RegistrationData&&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> >&&, content::ServiceWorkerDatabase::Status&&) () at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:599
#21 base::internal::Invoker<base::internal::BindState<base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>, content::ServiceWorkerDatabase::RegistrationData, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> >, content::ServiceWorkerDatabase::Status>, void ()>::RunImpl<base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>, std::tuple<content::ServiceWorkerDatabase::RegistrationData, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> >, content::ServiceWorkerDatabase::Status>, 0ul, 1ul, 2ul>(base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>&&, std::tuple<content::ServiceWorkerDatabase::RegistrationData, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> >, content::ServiceWorkerDatabase::Status>&&, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>) () at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:673
#22 base::internal::Invoker<base::internal::BindState<base::OnceCallback<void (content::ServiceWorkerDatabase::RegistrationData const&, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> > const&, content::ServiceWorkerDatabase::Status)>, content::ServiceWorkerDatabase::RegistrationData, std::vector<content::ServiceWorkerDatabase::ResourceRecord, std::allocator<content::ServiceWorkerDatabase::ResourceRecord> >, content::ServiceWorkerDatabase::Status>, void ()>::RunOnce(base::internal::BindStateBase*) () at ../../../../../qtwebengine/src/3rdparty/chromium/base/bind_internal.h:641
#23 0x00007fffebb24c84 in base::OnceCallback<void ()>::Run() && () at ../../../../../qtwebengine/src/3rdparty/chromium/base/callback.h:98
#24 base::TaskAnnotator::RunTask () at ./../../../../../qtwebengine/src/3rdparty/chromium/base/task/common/task_annotator.cc:142
#25 0x00007fffebb3718c in base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl () at ./../../../../../qtwebengine/src/3rdparty/chromium/base/task/sequence_manager/thread_controller_with_message_pump_impl.cc:365
#26 0x00007fffebb376eb in base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoSomeWork () at ./../../../../../qtwebengine/src/3rdparty/chromium/base/task/sequence_manager/thread_controller_with_message_pump_impl.cc:219
#27 0x00007fffebb8e1b6 in base::MessagePumpLibevent::Run () at ./../../../../../qtwebengine/src/3rdparty/chromium/base/message_loop/message_pump_libevent.cc:208
#28 0x00007fffebb2627f in base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run () at ./../../../../../qtwebengine/src/3rdparty/chromium/base/task/sequence_manager/thread_controller_with_message_pump_impl.cc:463
#29 0x00007fffebb14439 in base::RunLoop::RunWithTimeout () at ./../../../../../qtwebengine/src/3rdparty/chromium/base/run_loop.cc:157
#30 0x00007fffeed2d374 in content::BrowserProcessSubThread::IOThreadRun () at ../../../../../qtwebengine/src/3rdparty/chromium/content/browser/browser_process_sub_thread.cc:158
#31 0x00007fffebb54238 in base::Thread::ThreadMain () at ./../../../../../qtwebengine/src/3rdparty/chromium/base/threading/thread.cc:312
#32 0x00007fffebb8898a in ThreadFunc () at ./../../../../../qtwebengine/src/3rdparty/chromium/base/threading/platform_thread_posix.cc:81
#33 0x00007ffff7da1f27 in start_thread (arg=<optimized out>) at pthread_create.c:479
#34 0x00007ffff7eb72ef in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
</details>

**Comment by The-Compiler** [comment]
Gaaah, so I have a 5.15 developer build now, and I can *not* reproduce the epic games issue above... I'll build 5.14 as well overnight, and then (hopefully) be able to open another Qt report with @toofar's information above (thanks for that!)

I really wish this issue wouldn't play hide-and-seek...

**Comment by toofar** [comment]
I forgot I had a (old, non-debug) 5.15 build laying around, I also can't make it crash on the sites linked here (there was one from leveldb that that may have been due to jumping between version with the same basedir).
Another data point for 5.14, on that epic site if I open the developer tools and go to Application > Service Workers and check "Update on reload" and reload the page I get the same backtrace.

**Comment by The-Compiler** [comment]
- Are there people who can reproduce this on Qt 5.15 on Archlinux? If so, can you please install my Qt debug packages and [get a stacktrace](https://github.com/qutebrowser/qutebrowser/blob/master/doc/stacktrace.asciidoc)? I remain unable to reproduce with Qt 5.15, and my [earlier simplebrowser reproducer](https://github.com/qutebrowser/qutebrowser/issues/5279#issuecomment-636807190) doesn't seem to crash anymore either...

- @toofar do you happen to have a Qt 5.14 release build with debugging symbols around as well? I wonder what the "real" stacktrace is, without those debug assertions happening.

**Comment by toofar** [comment]
That first stack trace I put on there was from an unstripped release build I think, so it had symbols for functions and such but argument names had been optimised out. I had tried to get a release build with separate debug information but it didn't work and I only realised it resulted in unstripped libs when looking at this. I don't have that build around anymore but I can make a new one if you want?
My guess is that a service worker (or some part of it) is not being serialized in the right state. Then either it is being re-registered or updated when the site is loaded again and that code trips over the bad state. So if you don't use a session it won't happen when you start up but when you visit the site again. I don't know what state the code is expecting though or when it is being serialized from or how it could have anything to do with Qt. Or why it is so hard to reproduce with demo/example sites.

**Comment by The-Compiler** [comment]
Ah, I missed that your stack trace above already is from a debug build but not a developer build (i.e. with assertions). I have one myself now, and indeed it [segfaults in the same way](https://paste.the-compiler.org/view/aa73d9bb) yours does.

With the developer build, I get [the same debug assertion](https://paste.the-compiler.org/view/44b14d74) you already mentioned: `DCHECK(!handle_.pending_association());`. According to a QtWebEngine developer, the shutdown backtrace is likely unrelated.

So, some open questions:

- For Qt 5.14, why does this happen for qutebrowser but seemingly not (or not always) with other projects using QtWebEngine? What is it that we do differently?!
- Is the thing that some people report for Qt 5.15 really the same issue than Qt 5.14? Can someone show a stacktrace? I'm in contact with ~10 people so far, with no luck yet...
- Should we indeed just clean out the service worker directory between restarts as a workaround for this? Only with Qt 5.14, or also with Qt 5.15?

**Comment by toofar** [comment]
Here is a reproducer for 5.14, seems to be something to do with having both a service worker and a shared worker and re-registering them. empty1.js and empty2.js are empty files, but they don't need to be (I started out with the two from the epic store). You don't even need to start the shared worker.

```html
<html>
    <script>
      navigator.serviceWorker.register("/empty1.js")

      new SharedWorker('/empty2.js');
    </script>
</html>
```

**Comment by The-Compiler** [comment]
@toofar Thanks for that awesome investigation :heart: It looks like it's too late to get a fix for Qt 5.14, and at this point I'm not sure if what we're seeing on Qt 5.15 is in fact the same bug (it looks like that's *not* the case?).

Thus, I'd like to find a workaround for Qt 5.14 users. Indeed nuking the service worker directory might be a solution, but it's a quite aggressive one (and also somewhat time-consuming if you need to delete what might be dozens/hundreds of MBs before starting).

Interestingly, running with `--qt-flag disable-shared-workers` fixes the crash with your minimal example ("Uncaught ReferenceError: SharedWorker is not defined"), but not with the epic store website. Do you have any idea why that might be?

**Comment by toofar** [comment]
The original source has try/catches around both steps so you wouldn't have got that error there. But that flag makes it not crash for me on `https://www.epicgames.com/store/en-US/product/old-world/home` (after deleting `data/webengine/`, starting up and loading that page, :wq, then open again, on both a release and debug build of 5.14.2).

**Comment by The-Compiler** [comment]
Hm, right - not sure what I was testing wrong, but indeed `--qt-flag disable-shared-workers` *does* help. I just pushed a commit which adds that automatically if on Qt 5.14. I'm not sure yet what's going on exactly on Qt 5.15, but it does look like it's not really this issue.

Thanks again everyone for testing, and @toofar for digging into the details!

**Comment by The-Compiler** [comment]
It seems like this still *is* around in some way with Qt 5.15, at least on Windows. Unfortunately, the reproducer from above does not trigger the crash there, and apparently the workaround does not help either... I opened #5634 to continue tracking this.
